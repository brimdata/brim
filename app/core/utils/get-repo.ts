import packageJSON from "./package-json"

export async function getRepo(url?: string) {
  url = url || (await packageJSON()).repository
  return new URL(url).pathname.slice(1)
}
