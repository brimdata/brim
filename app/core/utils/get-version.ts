import {remote} from "electron"

export function getVersion() {
  return remote.app.getVersion()
}
