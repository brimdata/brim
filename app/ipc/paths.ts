import {app} from "electron"
import os from "os"
import {join} from "path"
import {createClient} from "src/pkg/electron-ipc-service"

const exe = (name) => (os.platform() === "win32" ? `${name}.exe` : name)

class Paths {
  brimcap() {
    return join(
      app.getAppPath().replace("app.asar", "app.asar.unpacked"),
      "zdeps",
      exe("brimcap")
    )
  }
}

export const paths = new Paths()
export const pathsClient = createClient(Paths)
