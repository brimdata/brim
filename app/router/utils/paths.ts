export const workspacePath = (id) => `/workspaces/${id}`
export const workspacesPath = () => `/workspaces`
export const lakePath = (id, workspaceId) =>
  `/workspaces/${workspaceId}/lakes/${id}`
