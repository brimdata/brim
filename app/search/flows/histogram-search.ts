import {Thunk} from "../../../src/js/state/types"
import {createSearchEntry} from "../../../src/js/brim/searchEntry"
import {search} from "../../../src/js/flows/search/mod"
import Current from "../../../src/js/state/Current"
import ErrorFactory from "../../../src/js/models/ErrorFactory"
import Last from "../../../src/js/state/Last"
import Notice from "../../../src/js/state/Notice"
import Search from "../../../src/js/state/Search"
import Tabs from "../../../src/js/state/Tabs"
import chart from "../../../src/js/state/Chart"
import {SearchResponse} from "../../../src/js/flows/search/response"

type Args = {
  query: string
  from: Date
  to: Date
}

const id = "Histogram"

export function histogramSearch({query, from, to}: Args): Thunk<Promise<void>> {
  return (dispatch, getState) => {
    const state = getState()

    const spaceId = Current.mustGetSpace(state).id
    const {response, promise} = dispatch(
      search({id, query, from, to, spaceId, target: "events"})
    )
    dispatch(handle(response))
    return promise
  }
}

function handle(response: SearchResponse): Thunk {
  return function(dispatch, getState) {
    const tabId = Tabs.getActive(getState())
    const current = Search.getCurrentRecord(getState())
    const previous = Last.getSearch(getState())

    if (shouldClear(current, previous)) dispatch(chart.clear(tabId))
    dispatch(chart.setStatus(tabId, "FETCHING"))

    response
      .status((status) => dispatch(chart.setStatus(tabId, status)))
      .chan(0, (records) => dispatch(chart.appendRecords(tabId, records)))
      .error((error) => dispatch(Notice.set(ErrorFactory.create(error))))
  }
}

function shouldClear(curr, prev) {
  if (!prev) return true

  const a = createSearchEntry(curr)
  const b = createSearchEntry(prev)

  if (
    a.program === b.program &&
    a.spaceId === b.spaceId &&
    a.getDuration() === b.getDuration()
  ) {
    return false
  } else {
    return true
  }
}
