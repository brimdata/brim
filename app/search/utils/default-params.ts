import {BrimSpace} from "src/js/brim"
import {SpanArgs} from "src/js/state/Search/types"

export function mergeDefaultSpanArgs(
  spanArgs: Partial<SpanArgs>,
  space: BrimSpace
): SpanArgs {
  const [from, to] = spanArgs
  if (!from && !to) return space.everythingSpan()
  if (!from) return [space.minTs(), to]
  if (!to) return [from, space.maxTs()]
  return [from, to]
}
