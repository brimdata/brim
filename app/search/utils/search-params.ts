import {TimeArg, SpanArgs} from "src/js/state/Search/types"

type Args = {
  program: string
  spanArgs: SpanArgs
  pins: string[]
  spanArgsFocus: SpanArgs
}

export const encodeSearchParams = ({
  program,
  pins,
  spanArgs,
  spanArgsFocus
}: Partial<Args>) => {
  const p = new URLSearchParams()
  if (program) p.append("q", program)
  encodeSpan(p, spanArgs, "from", "to")
  encodeSpan(p, spanArgsFocus, "focusFrom", "focusTo")
  encodePins(p, pins || [])
  return p.toString()
}

type DecodedSearchParams = {
  program: string
  pins: string[]
  spanArgs: Partial<SpanArgs>
  spanArgsFocus: Partial<SpanArgs>
}

export const decodeSearchParams = (path: string): DecodedSearchParams => {
  const url = new URLSearchParams(path)
  return {
    program: url.get("q"),
    spanArgs: [url.get("from"), url.get("to")].map(decodeSpanArg) as SpanArgs,
    spanArgsFocus: [url.get("focusFrom"), url.get("focusTo")].map(
      decodeSpanArg
    ) as SpanArgs,
    pins: decodePins(url)
  }
}

const encodeSpan = (params, span, from, to) => {
  if (span && span[0]) params.append(from, encodeSpanArg(span[0]))
  if (span && span[1]) params.append(to, encodeSpanArg(span[1]))
}

const encodeSpanArg = (arg) => {
  return typeof arg === "string" ? arg : `${arg.sec}.${arg.ns}`
}

const decodeSpanArg = (arg): TimeArg | null => {
  if (!arg) return null
  if (/^\d+\.\d+$/.test(arg)) {
    const [sec, ns] = arg.split(".").map((i) => parseInt(i))
    return {sec, ns}
  } else {
    return arg
  }
}

const pinKey = (i) => `p${i}`

const encodePins = (params, pins) => {
  for (let i = 0; i < pins.length; ++i) {
    params.append(pinKey(i), pins[i])
  }
}

const decodePins = (params) => {
  const pins = []
  for (let i = 0; params.has(pinKey(i)); i++) pins.push(params.get(pinKey(i)))

  return pins
}
