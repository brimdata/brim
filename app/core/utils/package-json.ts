import {pathsClient} from "app/ipc/paths"
import {readJSON} from "fs-extra"
import {join} from "path"

let cache

export default async function packageJSON() {
  if (cache) return cache
  const root = await pathsClient.root()
  const filePath = join(root, "package.json")
  return (cache = readJSON(filePath))
}
