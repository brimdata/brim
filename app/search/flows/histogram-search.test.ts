import {lakePath} from "app/router/utils/paths"
import Chart from "src/js/state/Chart"
import Current from "src/js/state/Current"
import Handlers from "src/js/state/Handlers"
import Spaces from "src/js/state/Spaces"
import Workspaces from "src/js/state/Workspaces"
import fixtures from "src/js/test/fixtures"
import initTestStore from "src/js/test/initTestStore"
import responses from "src/js/test/responses"
import {createZealotMock} from "zealot"
import {histogramSearch} from "./histogram-search"

const countByPathResp = responses("count_by_path.txt")
const space = fixtures("space1")

let store, zealot, dispatch, select
beforeEach(() => {
  zealot = createZealotMock()
  store = initTestStore(zealot.zealot)
  dispatch = store.dispatch
  select = (s: any) => s(store.getState())

  store.dispatchAll([
    Workspaces.add({
      host: "testHost",
      id: "1",
      name: "testName",
      port: "9867",
      authType: "none"
    }),
    Current.setWorkspaceId("1"),
    Spaces.setDetail("1", space),
    Current.setSpaceId(space.id)
  ])
  global.tabHistory.push(lakePath(space.id, "1"))
  zealot.stubStream("search", countByPathResp)
})

const submit = () =>
  dispatch(histogramSearch({query: "dns query | every 1s count() by _path"}))

test("zealot gets the request", async () => {
  await submit()
  const calls = zealot.calls("search")
  expect(calls.length).toBe(1)
  expect(calls[0].args).toEqual("dns query | every 1s count() by _path")
})

test("the chart status updates", async () => {
  const promise = submit()
  expect(select(Chart.getStatus)).toBe("FETCHING")
  await promise
  expect(select(Chart.getStatus)).toBe("SUCCESS")
})

test("registers historgram request then cleans it up", async (done) => {
  const promise = submit()
  expect(select(Handlers.get)["Histogram"]).toEqual(
    expect.objectContaining({type: "SEARCH"})
  )
  await promise
  // The promise only waits for the table, might be good to return two
  // promises so people can decide what they want to wait for.
  setTimeout(() => {
    expect(select(Handlers.get)["Histogram"]).toBe(undefined)
    done()
  })
})

test("aborts previous histogram request", async () => {
  const abort = jest.fn()
  dispatch(Handlers.register("Histogram", {type: "SEARCH", abort}))
  await submit()
  expect(abort).toHaveBeenCalledTimes(1)
})

test("populates the chart", async () => {
  await submit()
  expect(select(Chart.getData)).toMatchSnapshot()
})
