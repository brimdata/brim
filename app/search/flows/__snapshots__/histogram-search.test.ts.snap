// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`populates the chart 1`] = `
Object {
  "keys": Array [
    "capture_loss",
    "conn",
    "files",
    "x509",
    "ssl",
    "dns",
    "weird",
    "stats",
  ],
  "table": Object {
    "1582646585000": Object {
      "conn": 1,
      "stats": 1,
    },
    "1582646586000": Object {
      "conn": 1,
      "weird": 1,
    },
    "1582646587000": Object {
      "conn": 2,
    },
    "1582646588000": Object {
      "conn": 5,
    },
    "1582646589000": Object {
      "conn": 1,
    },
    "1582646590000": Object {
      "conn": 2,
    },
    "1582646591000": Object {
      "conn": 1,
      "dns": 1,
    },
    "1582646592000": Object {
      "conn": 1,
    },
    "1582646593000": Object {
      "conn": 2,
      "dns": 1,
    },
    "1582646594000": Object {
      "files": 4,
      "ssl": 1,
      "x509": 1,
    },
    "1582646595000": Object {
      "conn": 3,
    },
    "1582646597000": Object {
      "capture_loss": 1,
    },
  },
}
`;
